src/Raven.Server/Web/System/AdminGcDebugHandler.cs

```diff
@@ -0,0 +1,424 @@
+                        var generation = (uint)eventData.Payload[1];
+                        var reason = (uint)eventData.Payload[2];
+                        timeGCStartByIndex[startIndex] = (eventData.TimeStamp, generation, reason);
+                        break;
+
+                    case "GCEnd_V1":
+                        var endIndex = long.Parse(eventData.Payload[0].ToString());
+
+                        if (timeGCStartByIndex.TryGetValue(endIndex, out var tuple) == false)
+                            return;
+
+                        _events.Add(new GCEvent(tuple.DateTime, eventData, endIndex, tuple.Generation, tuple.Reason));
+                        timeGCStartByIndex.Remove(endIndex);
+                        break;
+
+                    case "GCSuspendEEBegin_V1":
+                        _suspendData = eventData;
+                        break;
+
+                    case "GCSuspendEEEnd_V1":
+                        if (_suspendData == null)
+                        {
+                            Console.WriteLine($"WHAT???");
+                            return;
+                        }
+
+                        var index = (uint)_suspendData.Payload[1];
+                        var suspendReason = (uint)_suspendData.Payload[0];
+
+                        _events.Add(new GCSuspendEvent(_suspendData.TimeStamp, eventData, index, suspendReason));
+                        _suspendData = null;
+                        break;
+
+                    case "GCRestartEEBegin_V1":
+                        timeGCRestartStart = eventData.TimeStamp;
+                        break;
+
+                    case "GCRestartEEEnd_V1":
+                        if (timeGCRestartStart == null)
+                            return;
+
+                        _events.Add(new Event(EventType.GCRestart, timeGCRestartStart.Value, eventData));
+                        timeGCRestartStart = null;
+                        break;
+
+                    case "GCFinalizersBegin_V1":
+                        timeGCFinalizersStart = eventData.TimeStamp;
+                        break;
+
+                    case "GCFinalizersEnd_V1":
+                        if (timeGCFinalizersStart == null)
+                            return;
+
+                        _events.Add(new Event(EventType.GCFinalizers, timeGCFinalizersStart.Value, eventData));
+                        timeGCFinalizersStart = null;
+                        break;
+                }
+            }
+        }
+
+        internal abstract class Expensive_GcEventListener : EventListener
+        {
+            private const int GC_KEYWORD = 0x0000001;
+
+            protected override void OnEventSourceCreated(EventSource eventSource)
+            {
+                if (eventSource.Name.Equals("Microsoft-Windows-DotNETRuntime"))
+                {
+                    EnableEvents(eventSource, EventLevel.Verbose, (EventKeywords)GC_KEYWORD);
```

Comment:
should we disable them? or this is handled internally?

Reply:
it should be handled internally on dispose. I added it anyway.